src/providers/postgres/qgspostgresconn.cpp

```diff
@@ -1924,7 +1924,17 @@ void QgsPostgresConn::retrieveLayerTypes( QVector<QgsPostgresLayerProperty *> &l
 
       sql += QLatin1String( ") " );
 
-      sql += " FROM " + table;
+      if ( type == QgsWkbTypes::Unknown )
```

Comment:
this looks a bit convoluted to me. There might perhaps be a good reason, but why not just appending ``QStringLiteral(" LIMIT %1").arg( GEOM_TYPE_SELECT_LIMIT )`` after the existing ``sql += " FROM " + table;`` ?
I'm not clear why the restriction is only when the layer geometry type is unknown. Wouldn't the issue also occur with big tables of known geometry type ?

Reply:
The point is:

https://github.com/qgis/QGIS/blob/6ae5d22f1da6ddf8c518919d7b5b36cfad5a3d3e/src/providers/postgres/qgspostgresconn.cpp#L1893

and
https://github.com/qgis/QGIS/blob/6ae5d22f1da6ddf8c518919d7b5b36cfad5a3d3e/src/providers/postgres/qgspostgresconn.cpp#L1912

https://github.com/qgis/QGIS/blob/6ae5d22f1da6ddf8c518919d7b5b36cfad5a3d3e/src/providers/postgres/qgspostgresconn.cpp#L1914

Only for unknown geometry the `array_agg` is performed on the WHOLE table for `UPPER(geometry)`, else its performed on static texts.
We would like to apply the `limit` to the `array_agg` but SQL does not allow this.

Appending the LIMIT would still query the whole table.

Reply:
thanks for the explanations. Maybe adding a bit of comment in the code would help future readers

Reply:
@rouault Yes, I will add some comments